pkg/metrics/metrics.go

```diff
@@ -0,0 +1,189 @@
+	prometheus.MustRegister(MachineSetStatusAvailableReplicas)
+	prometheus.MustRegister(MachineSetStatusReadyReplicas)
+	prometheus.MustRegister(MachineSetStatusReplicas)
+}
+
+type MachineCollector struct {
+	machineLister    machinelisters.MachineLister
+	machineSetLister machinelisters.MachineSetLister
+	namespace        string
+}
+
+func NewMachineCollector(machineInformer machineinformers.MachineInformer, machinesetInformer machineinformers.MachineSetInformer, namespace string) (*MachineCollector, error) {
+	return &MachineCollector{
+		machineLister:    machineInformer.Lister(),
+		machineSetLister: machinesetInformer.Lister(),
+		namespace:        namespace,
+	}, nil
+}
+
+// Collect is method required to implement the prometheus.Collect interface.
+func (mc *MachineCollector) Collect(ch chan<- prometheus.Metric) {
+	mc.collectMachineMetrics(ch)
+	mc.collectMachineSetMetrics(ch)
+}
+
+// Describe implements the prometheus.Collector interface.
+func (mc MachineCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- MachineCountDesc
+	ch <- MachineSetCountDesc
+}
+
+// Collect implements the prometheus.Collector interface.
+func (mc MachineCollector) collectMachineMetrics(ch chan<- prometheus.Metric) {
+	machineList, err := mc.listMachines()
+	if err != nil {
+		ScrapeFailedCounter.With(prometheus.Labels{"kind": "Machine-count", "reason": err.Error()}).Inc()
+		return
+	}
+
+	for _, machine := range machineList {
+		mMeta := machine.ObjectMeta
+		typeMeta := machine.TypeMeta
+		mSpec := machine.Spec
+		providerid := ""
+		if mSpec.ProviderID != nil {
+			providerid = *mSpec.ProviderID
+		}
+
+		MachineInfo.With(prometheus.Labels{
+			"name":             mMeta.Name,
+			"namespace":        mMeta.Namespace,
+			"createdAt":        strconv.FormatInt(mMeta.GetCreationTimestamp().Time.Unix(), 10),
+			"spec_provider_id": providerid,
+			"api_version":      typeMeta.APIVersion,
+			"node":             machine.Status.NodeRef.Name}).Set(float64(1))
+	}
+
+	metric, err := prometheus.NewConstMetric(MachineCountDesc, prometheus.GaugeValue, float64(len(machineList)))
+	if err != nil {
+		ScrapeFailedCounter.With(prometheus.Labels{"kind": "Machine-count", "reason": err.Error()}).Inc()
+		return
+	}
+	ch <- metric
+	glog.V(4).Infof("collectmachineMetrics exit")
+
+}
+
+// collectMachineSetMetrics is method to collect machineSet related metrics.
+func (mc MachineCollector) collectMachineSetMetrics(ch chan<- prometheus.Metric) {
+	machineSetList, err := mc.listMachineSets()
+	if err != nil {
+		ScrapeFailedCounter.With(prometheus.Labels{"kind": "Machineset-count", "reason": err.Error()}).Inc()
+		return
+	}
+	metric, err := prometheus.NewConstMetric(MachineSetCountDesc, prometheus.GaugeValue, float64(len(machineSetList)))
+	if err != nil {
+		ScrapeFailedCounter.With(prometheus.Labels{"kind": "Machineset-count", "reason": err.Error()}).Inc()
+		return
+	}
+	ch <- metric
+
+	for _, machineSet := range machineSetList {
```

Comment:
I think it is easier to read without the temporaries: 

```go
	for _, machineSet := range machineSetList {
		MachineSetInfo.With(prometheus.Labels{
			"name":        machineSet.Name,
			"namespace":   machineSet.Namespace,
			"createdAt":   strconv.FormatInt(machineSet.GetCreationTimestamp().Time.Unix(), 10),
			"api_version": machineSet.APIVersion}).Set(float64(1))

		MachineSetInfoSpecReplicas.With(prometheus.Labels{
			"name":      machineSet.Name,
			"namespace": machineSet.Namespace}).Set(float64(pointer.Int32PtrDerefOr(machineSet.Spec.Replicas, 0)))

		MachineSetStatusAvailableReplicas.With(prometheus.Labels{
			"name":      machineSet.Name,
			"namespace": machineSet.Namespace,
		}).Set(float64(machineSet.Status.AvailableReplicas))

		MachineSetStatusReadyReplicas.With(prometheus.Labels{
			"name":      machineSet.Name,
			"namespace": machineSet.Namespace,
		}).Set(float64(machineSet.Status.ReadyReplicas))

		MachineSetStatusReplicas.With(prometheus.Labels{
			"name":      machineSet.Name,
			"namespace": machineSet.Namespace,
		}).Set(float64(machineSet.Status.ReadyReplicas))
	}
```

Also note the safe deref of `Replicas` in this ^^ diff: 

```go
		MachineSetInfoSpecReplicas.With(prometheus.Labels{
			"name":      machineSet.Name,
			"namespace": machineSet.Namespace}).Set(float64(pointer.Int32PtrDerefOr(machineSet.Spec.Replicas, 0)))
```